conn.go

```diff
@@ -69,13 +69,19 @@ type Conn struct {
 	writeHeaderBuf [8]byte
 	writeHeader    header
 
+	// Close handshake state.
+	closeStateMu     sync.RWMutex
+	closeReceivedErr error
+	closeSentErr     error
+
+	// CloseRead state.
 	closeReadMu   sync.Mutex
 	closeReadCtx  context.Context
 	closeReadDone chan struct{}
 
-	closed  chan struct{}
+	closing atomic.Bool
```

Comment:
Would it be more idiomatic to have `closing chan struct{}` that we close to indicate that the connection has entered the "closing" state?

Reply:
I wouldn't say more idiomatic per-se, but at least here interacting with the channel would be racy, it would require protection via mutex which I wanted to avoid. See my other comment for a motivation for this.